File: src/network/packets/playerState.js
import Packet from '../packet.js';
import Writer from '../utils/writer.js';
import Player from "../../ui/objects/player.js";
import {convertIntColorToHex} from "../../ui/utils.js";
import Point from "../../ui/objects/point";


class PlayerStatePacket extends Packet {

    constructor(userId, mapSize) {
        super();
        this.userId = userId;
        this.packetId = 1004;
        this.player = null;

    }


    // Handel Server Response
    parsePacket() {
        const reader = this.reader;
        this.userId = reader.readInt4();
        this.playerName = reader.readString();

        this.playerX = reader.readInt2();
        this.playerY = reader.readInt2();
        this.direction = reader.readString();


        // Colors
        this.colorBrighter = convertIntColorToHex(reader.readInt4());
        this.colorDarker = convertIntColorToHex(reader.readInt4());
        this.colorSlightlyBrighter = convertIntColorToHex(reader.readInt4());
        this.colorPattern = convertIntColorToHex(reader.readInt4());
        this.colorPatternEdge = convertIntColorToHex(reader.readInt4());


    }

    finalize() {
        // Handle Server Request
        // Send Empty Packet As Ask For Ready
        const writer = new Writer(this.packetId);
        return writer.finalize();
    }


    handleReceivedPacket(client) {

        const myPlayer = client.player;

        let player = new Player(new Point(0, 0), this.userId);
        player = window.gameEngine.gameObjects.addPlayer(player);

        player.name = this.playerName;
        player.colorBrighter = this.colorBrighter;
        player.colorDarker = this.colorDarker;
        player.colorSlightlyBrighter = this.colorSlightlyBrighter;
        player.colorPattern = this.colorPattern;
        player.colorPatternEdge = this.colorPatternEdge;




        // When Receiving Player State
        // Next Frame Move Relative To Server Pos
        player.hasReceivedPosition = true;
        player.moveRelativeToServerPosNextFrame = true;
        player.lastServerPosSentTime = Date.now();

        // current player consider that his last position has been confirmed
        myPlayer.lastPosHasBeenConfirmed = true;


        let offset = player.calMoveOffset();
        let newPos = new Point(this.playerX, this.playerY);
        let newPosOffset = newPos.clone();
        let newDir = this.direction;

        newPosOffset = Player.movePlayer(newPosOffset, newDir, offset);
        let clientServerNeedsSync = true;



        if (player.isMyPlayer) {


            player.lastPosServerSentTime = Date.now();

            // Check If Server Synced With Client
            // To Draw This Movement or Ignore It
            // if server predict the same movement
            // or the movement is to close to server
            clientServerNeedsSync = player.checkClientMovementSyncedWithServer(newDir
                , newPosOffset, newPos);

            if (clientServerNeedsSync) {
                /***
                 Here We Found That Server and Client not Synced
                 So We Need To Sync Them
                 1- Change Player Direction
                 2- Change Player Position
                 3- Request Waiting Blocks From Server
                 4- Clear Send Dir Queue
                 */
                player.changeCurrentDir(newDir, newPos, false, false);
                player.requestWaitingBlocks();
                player.sendDirQueue = [];
            }

            player.serverPos = newPosOffset.clone();
            player.serverDir = newDir;

            player.removeBlocksOutsideCamera();
        } else {
            player.updatePlayerDirection(newDir);
        }

        if (clientServerNeedsSync) {
            player.position = newPosOffset.clone();
            player.addWaitingBlocks(newPos);
        }

        //Start To Handel Draw Position
        if (!player.drawPosSet) {
            // if we don't draw this player before set draw position
            player.drawPosSet = true;
            player.drawPosition = player.position.clone();
        }


    }
}

export default PlayerStatePacket;

File: gameserver/network/packets/player_state.py
from gameserver.network.packet import Packet
from gameserver.network.utils.writer import Writer


class PlayerStatePacket(Packet):
    PACKET_ID = 1004
    def __init__(self, player):
        super().__init__()
        self.packet_id = self.PACKET_ID
        self.user_id = player.player_id
        self.player_name = player.name
        self.player_x = player.position.x
        self.player_y = player.position.y
        self.player_direction = player.direction

        ## COLORS
        self.color_brighter = player.color_brighter
        self.color_darker = player.color_darker
        self.color_slightlyBrighter = player.color_slightlyBrighter
        self.color_pattern = player.color_pattern
        self.color_patternEdge = player.color_patternEdge

    def parse_packet(self):
       pass

    def handle_packet(self, client):
        """On Received Ready Packet"""
        print(client.player.name, "is Recived Player State Packet")

    def finalize(self):
        writer = Writer(self.packet_id)
        writer.write_int_in_bytes(self.user_id, 4)
        writer.write_string(self.player_name)

        # Send Player Position
        writer.write_int_in_bytes(self.player_x, 2)
        writer.write_int_in_bytes(self.player_y, 2)
        writer.write_string(self.player_direction)

        # Send Player Colors
        writer.write_int_in_bytes(self.color_brighter, 4)
        writer.write_int_in_bytes(self.color_darker, 4)
        writer.write_int_in_bytes(self.color_slightlyBrighter, 4)
        writer.write_int_in_bytes(self.color_pattern, 4)
        writer.write_int_in_bytes(self.color_patternEdge, 4)

        return writer.finalize()
